build: complete user flow; vendor and vendor_repo

caramandvd opened the pull request.

caramandvd committed:
refactor: imported localy the engine from Utils.util.py instead of creating it locally every time

caramandvd committed:
refactor: added relationship key

caramandvd committed:
bulid: requirements.txt

caramandvd committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	model/domain/currencies.py
#	model/repository/atm_locations_repository.py
#	model/repository/messages_repository.py
#	model/repository/users_credentials_repository.py
#	model/repository/users_repository.py
#	model/repository/users_transaction_repository.py

caramandvd committed:
bulid: created Flask server; users_controller.py

caramandvd committed:
bulid: created Flask server; users_controller.py

caramandvd committed:
bulid:  users_credentials_repository.py_controller.py; added Error handling for unexisting user

caramandvd committed:
bulid: Complete user flow. AppController. server.py

caramandvd committed:
bulid: added vendors_repository.py and vendors.py. Changed in users_transaction.py from vendor_id to reciever_id

EcclesiaITAcademy commented on controller/app_controller.py:
```diff
@@ -0,0 +1,44 @@
+    def __init__(self, users_repository):
+        self.users_repository = users_repository
+
+    def create_user(self, user_id, first_name, last_name, email, address, phone_number):
```
Are we ever going to use this? We use add_user all the time, no? We'll never create a user without a username and password

EcclesiaITAcademy commented on controller/app_controller.py:
```diff
@@ -0,0 +1,44 @@
+    def get_user_by_phone_nr(self, phone_number):
+        return UsersRepository().read_user_by_phone_number(phone_number=phone_number)
+
+    def add_username_password(self, user_id, username, user_password_hash):
```
Same here as above

EcclesiaITAcademy commented on controller/users_controller.py:
```diff
@@ -0,0 +1,32 @@
+from model.repository.users_repository import UsersRepository
```
This file is no longer needed since we'll go directly from AppController to the Repositories

EcclesiaITAcademy commented on controller/users_credentials_controller.py:
```diff
@@ -0,0 +1,27 @@
+from model.repository.users_credentials_repository import UsersCredentialsRepository
```
This file is no longer needed since we'll go directly from AppController to the Repositories

EcclesiaITAcademy commented on model/domain/users_transaction.py:
```diff
@@ -10,7 +10,7 @@ class UsersTransactions(Base):
     currency = Column(String(3), foreign_key=ForeignKey("currencies.currency", ondelete="CASCADE"))
     amount = Column(Float(2))
-    vendor = Column(String(100))
+    reciever_id = Column(String(100))
```
receiver_id

EcclesiaITAcademy commented on server/server.py:
```diff
@@ -0,0 +1,35 @@
+
+@app.route('/api/v1/create_user', methods=['POST'])
+def create_user():
+    if request.method == 'POST':
```
I will always be a post method. using get on this endpoint should fail

EcclesiaITAcademy commented on server/server.py:
```diff
@@ -0,0 +1,35 @@
+def create_user():
+    if request.method == 'POST':
+        # date = request.get_json()
+        user_id = request.json['user_id']
```
You can put the requests.json value in another variable. It looks cleaner and you don't have to rewrite request.json over and over again

EcclesiaITAcademy commented on server/server.py:
```diff
@@ -0,0 +1,35 @@
+        username = request.json['username']
+        user_password_hash = request.json['password']
+
+        AppController(UsersRepository()).add_user(user_id=user_id,
```
Create an AppController variable that you will use in the whole server. 

user_repository = UserRepository()
app_ctrl = AppController(user_repository)

And use it like:
app_ctrl.add_user(...)

EcclesiaITAcademy commented on server/server.py:
```diff
@@ -0,0 +1,35 @@
+                               user_password_hash=user_password_hash)
+
+        return jsonify({
+            'response': "User: " + user_id + "; " + first_name + "; " + last_name + "; " + email + "; " + address + "; "
```
Don't return a text. Return it like a dict (or json):
{"response": {
    "user_id" : user_id,
    "first_name": first_name
    ... 
}

That way when the data gets to the UI, it does not need to parse a string, but rather use a json file to get the data. It's easier access

caramandvd committed:
build: "get previous transaction" flow completed;
build: read_all_transactions_by_id method in users_transaction_repository.py
style: typos 'reciever_id' to 'receiver_id'
